### PR TITLE
TOP - Tic Tac Toe

### PR DESCRIPTION
**Pull request by kcisneros**
Working through the Tic Tac Toe game assignment from the Odin project. https://www.theodinproject.com/lessons/ruby-tic-tac-toe#assignment

### CODE REVIEW

**Comment by kcisneros** [comment]
I tried thinking about what sort of tests I'd add here as I've done in previous exercises, and I'm not sure what they would be.. I tested different scenarios as I went along, but obviously that's not fool-proof. While testing isn't required during the odin project exercises, at least at this stage. I was trying to build that habit. While I'm happy with the final result and understand all of the code, I was not able to get here on my own. A testable version can be seen here: https://replit.com/@kcisnerosreplit/SardonicScornfulSystem

File: ruby_oop/tic_tac_toe/game.rb [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+       (@board.board[3] == 'X' && @board.board[4] == 'X' && @board.board[5] == 'X') || # second row horizontal
+       (@board.board[3] == 'O' && @board.board[4] == 'O' && @board.board[5] == 'O') || # second row horizontal
+       (@board.board[6] == 'X' && @board.board[7] == 'X' && @board.board[8] == 'X') || # third row horizontal
+       (@board.board[6] == 'O' && @board.board[7] == 'O' && @board.board[8] == 'O') || # third row horizontal
+       (@board.board[1] == 'X' && @board.board[4] == 'X' && @board.board[7] == 'X') || # second row vertical
+       (@board.board[1] == 'O' && @board.board[4] == 'O' && @board.board[7] == 'O') || # second row vertical
+       (@board.board[2] == 'X' && @board.board[5] == 'X' && @board.board[8] == 'X') || # third row vertical
+       (@board.board[2] == 'O' && @board.board[5] == 'O' && @board.board[8] == 'O') || # third row vertical
+       (@board.board[2] == 'X' && @board.board[4] == 'X' && @board.board[6] == 'X') || # top right diagonal
+       (@board.board[2] == 'O' && @board.board[4] == 'O' && @board.board[6] == 'O') # top right diagonal
```
Comment (zacharydanger) [inline ruby_oop/tic_tac_toe/game.rb:65]: Brute force is perfectly valid when you're figuring things out. 😀

Here are the steps I'd take to refactor when I see some obviously un-DRY.

Look for the actual _variables_ and try to factor them out, i.e. what's different from case to case. In this case I'm seeing two: `X vs O` and the indices of the winning vector.

I'll start with `X v O` because that seems more obvious. Do we care that these are specifically `X` or `O`? No. We really just care that they're all the same (and not `nil`).

```ruby
if (@board.board[0] == @board.board[1] && @board.board[0] == @board.board[2] && !@board.board[0].nil?) || # first row horizontal
```

This conditional works for both `X` and `O`. We don't have to compare _every_ position against every other position because of the _transitive properties of equality™_. (Also `board` is such a weird word when it's written out so many times o_0)

With that out of the way, we can start thinking about the winning vectors. Your instinct about using a constant is good. These appear to be just sets of 3 values `[0, 1, 2], [0, 3, 6]...` so let's throw those into an array:

```ruby
WINNING_VECTORS = [
  [0,1,2], # first row horizontal
  [0,3,6], # first row vertical
  # etc...
]
```

By looping through those, we can test each vector:

```ruby
if WINNING_VECTORS.any? { |v| @board.board[v[0]] == @board.board[v[1]] && @board.board[v[0]] == @board.board[v[2]] && @board.board[v[0]].nil? }
  game_over = true
  puts "Game over!'
elsif # tie condition, etc
```

There's still more refactoring that can happen, but holy crap I already got carried away.

**TLDR**: I'm down to pair on this if need-be. ¯\\\_(ツ)_/¯

File: ruby_oop/tic_tac_toe/board.rb [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+    # while @board[@spot].is_a? String
+    #   puts 'Already taken! Please pick another spot'
+    #   ask_for_placement
+    #   @board[@spot] = marker
+    #   break if @board[@spot].is_a? Integer
+    # end
+    # @board[@spot] = marker
+
+    if @board[@spot].is_a? Integer
```
Comment (mnishiguchi) [inline ruby_oop/tic_tac_toe/board.rb:43]: For matching Class, sometime the case-when statement is useful.

```rb
case @board[@spot]
  ...
when String
  ...
when Integer
  ...
else
  raise "invalid spot: #{@board[@spot]}"
end
```

Also if we want to check if the index is valid, we could use `fetch`. Which to use depends. Nice to have both in our toolbox.

```rb
> [1, 2, 3].fetch(1)
=> 2

> [1, 2, 3].fetch(9)
(irb):12:in `fetch': index 9 outside of array bounds: -3...3 (IndexError)
```

vs

```rb
> [1, 2, 3][1]
=> 2

> [1, 2, 3][9]
=> nil
```

File: ruby_oop/tic_tac_toe/game.rb (hunk 1 above)
Comment (chmurph2) [reply ruby_oop/tic_tac_toe/game.rb:65]: This is a good refactor. I also think it's good to have gone through the brute force approach to see the value in the refactor in the end. I remember going through something similar in my very first CS course in college, where my prof. basically gave me the same refactoring advice.

I recommend working through this refactor before getting back to the next assignment, as it's important.

File: ruby_oop/tic_tac_toe/game.rb [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        puts "#{player1.name.capitalize} is selecting a spot..."
+        @board.board_update(player1.marker)
+        turn_number += 1
+      else
+        puts "#{player2.name.capitalize} is selecting a spot..."
+        @board.board_update(player2.marker)
+        turn_number += 1
+      end
+    end
+  end
```
Comment (chmurph2) [inline ruby_oop/tic_tac_toe/game.rb:30]: I took a stab at refactoring `swap_players` into two methods, `play` and `swap_players`, in order to:

1. create a public method called `play` which I think reads better as `@game.play` vs. `@game.swap_players`. As a user of this class, I'm wondering why I need to call `swap_players` in order to "play" Tic Tac Toe.  Ya know?
2. DRYs up assigning `turn_number += 1`.
3. Makes `swap_players` private as that seems more like an internal concern of the class (see 1 above)

```suggestion
 
   def play
     turn_number = 1
     until game_over?
       swap_players(turn_number)
       turn_number += 1
     end
   end
 
   private
  
   def swap_players(turn_number)
     if turn_number.odd?
       puts "#{player1.name.capitalize} is selecting a spot..."
       @board.board_update(player1.marker)
     else
       puts "#{player2.name.capitalize} is selecting a spot..."
       @board.board_update(player2.marker)
     end
   end
```

What do you think? Happy to pair on this if I wasn't clear.

**Review by chmurph2 — APPROVED** [review]
